File: backend/duration.py
"""Duration parsing — one correct implementation, shared by the engines.

Downtime durations are free-text ("2 hrs 15 min", "1 hr 10 min", "45 min", a
bare "90", or a decimal "1.5 hrs"). This parser reads hours and minutes and sums
them to minutes. It used to be duplicated: analytics_engine had a correct regex
version, predictive_engine a cruder digit-concatenation version that misread any
hour format ("1 hr" -> 1 minute, "2 hrs 15 min" -> 215), which silently
understated downtime in the predictive-maintenance risk score. Both now import
this one function.

The number is matched as a decimal (\\d+(?:\\.\\d+)?) so "1.5 hrs" is 90 minutes,
not 300. An integer-only `\\d+` skipped the "1." and matched the "5" in "1.5 h",
reading a 90-minute stop as 5 hours — a 3x overstatement that flowed straight
into every downtime rollup, the cost-of-losses £ and the risk score. Operators
type this field free-hand, so decimals are real input.
"""
import re
from functools import lru_cache

_NUM = r"(\d+(?:\.\d+)?)"


# Memoised because the aggregate paths call this once per DISTINCT duration
# GROUP rather than once per row, and a factory types the same handful of
# durations over and over -- ~11,000 groups over ~55 distinct strings at 200
# machines with a month of history.
#
# Worth being precise about the size of this. I predicted it would remove most
# of /machine-health's Python time; it did not. Measured, it took that endpoint
# from 282.8 ms to 261.8 ms -- about 8%. The real cost was elsewhere (a
# projection scan in ai/twin.py). Kept because it is free and correct, not
# because it was the fix.
#
# Safe to cache: a pure function of its argument -- no clock, no database, no
# globals -- so the same string always yields the same number. Bounded at 4096
# because `value` is FREE TEXT an operator types, and an unbounded cache keyed
# on user input grows without limit.
@lru_cache(maxsize=4096)
def parse_duration_to_minutes(value: str) -> int:
    if not value:
        return 0

    lower = str(value).lower()
    total = 0.0

    hour_match = re.search(_NUM + r"\s*h", lower)
    minute_match = re.search(_NUM + r"\s*m", lower)

    if hour_match:
        total += float(hour_match.group(1)) * 60
    if minute_match:
        total += float(minute_match.group(1))

    # No unit at all — treat a bare number (integer or decimal) as minutes.
    if not hour_match and not minute_match:
        plain = re.search(_NUM, lower)
        total += float(plain.group(1)) if plain else 0.0

    return round(total)

File: frontend/lib/duration.ts
/**
 * Duration parsing — the frontend half of a pair.
 *
 * `downtime_logs.duration` is free text ("2 hrs 15 min", "45 min", a bare "90",
 * a decimal "1.5 hrs"). The backend reads it with
 * `backend/duration.py:parse_duration_to_minutes` to produce OEE, the
 * cost-of-losses figure and the risk scores. The dashboard reads the same rows
 * to render its Total Downtime tile. The two have to agree, so this is a
 * deliberate port of that function rather than an independent implementation —
 * `lib/duration.test.ts` pins it against the backend's actual output.
 *
 * The number is matched as a decimal, because an integer-only `\d+` skips the
 * "1." and matches the "5" in "1.5 h", reading a 90-minute stop as 5 hours.
 * That is the bug the backend fixed and this side had kept.
 */

const NUM = String.raw`(\d+(?:\.\d+)?)`;

const HOUR_RE = new RegExp(`${NUM}\\s*h`);
const MINUTE_RE = new RegExp(`${NUM}\\s*m`);
const PLAIN_RE = new RegExp(NUM);

// Python's round() breaks ties to even: round(2.5) is 2, round(3.5) is 4.
// Math.round would give 3 and 4. Matching it keeps the tile and the API on the
// same minute for every input, not just most of them.
//
// Exported because it is not a duration rule, it is the Python rule — anything
// this codebase ports from a `round(...)` in backend/ needs it. lib/shift.ts is
// the second caller: a shift that made 1 of a target 40 is exactly 2.5%, where
// Math.round says 3% and the API says 2%.
export function roundHalfToEven(n: number) {
  const floor = Math.floor(n);
  const remainder = n - floor;

  if (remainder > 0.5) return floor + 1;
  if (remainder < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}

export function parseDurationToMinutes(value: string | null | undefined) {
  if (!value) return 0;

  const lower = String(value).toLowerCase();
  let total = 0;

  const hourMatch = lower.match(HOUR_RE);
  const minuteMatch = lower.match(MINUTE_RE);

  if (hourMatch) total += Number(hourMatch[1]) * 60;
  if (minuteMatch) total += Number(minuteMatch[1]);

  // No unit at all — treat a bare number (integer or decimal) as minutes.
  if (!hourMatch && !minuteMatch) {
    const plain = lower.match(PLAIN_RE);
    total += plain ? Number(plain[1]) : 0;
  }

  return roundHalfToEven(total);
}
